<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from primo_real.xacro               | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="primo" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <!-- This script requires the alignment arguments to be sent in. They are
         used by the cameras xacro file -->
  <link name="base_link">
    <inertial>
      <mass value=".001"/>
      <origin rpy=" 0 0 0" xyz="0 0 0"/>
      <inertia ixx="1" ixy="0" ixz="1" iyy="1" iyz="0" izz="1"/>
    </inertial>
  </link>
  <joint name="base_joint" type="fixed">
    <axis xyz="0 0 0"/>
    <origin rpy="0 0 0" xyz="0.280 0 0"/>
    <parent link="base_link"/>
    <child link="chassis_link"/>
  </joint>
  <link name="chassis_link">
    <visual>
      <geometry>
        <!-- <mesh filename="package://primo_description/meshes/r1_primo_${type}.stl"/> -->
        <mesh filename="package://primo_description/meshes/primo.dae"/>
      </geometry>
    </visual>
    <collision name="collision">
      <origin xyz="-.274 0 .4"/>
      <geometry>
        <box size=".8 .66 .5"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="31.0"/>
      <origin rpy=" 0 0 0" xyz="-.274 0 .4"/>
      <inertia ixx="11.519" ixy="0" ixz="-3" iyy="14.163" iyz="0" izz="5"/>
    </inertial>
  </link>
  <link name="left_wheel">
    <visual>
      <geometry>
        <mesh filename="package://primo_description/meshes/r1_wheel.stl"/>
      </geometry>
    </visual>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <collision name="collision">
      <!-- <origin xyz="0 0 0" rpy="0 1.5707 1.5707"/> -->
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.1" radius="0.151"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <mass value="2"/>
      <inertia ixx=".1" ixy="0.0" ixz="0.0" iyy=".1" iyz="0.0" izz=".1"/>
    </inertial>
  </link>
  <joint name="left_wheel_joint" type="continuous">
    <parent link="chassis_link"/>
    <child link="left_wheel"/>
    <origin rpy="0 0 0" xyz="0 0.3175 0.1510"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
  </joint>
  <!-- Gazebo wheel reference -->
  <gazebo reference="left_wheel">
    <mu1 value="200.0"/>
    <mu2 value="100.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <material>Gazebo/Black</material>
  </gazebo>
  <link name="right_wheel">
    <visual>
      <geometry>
        <mesh filename="package://primo_description/meshes/r1_wheel.stl"/>
      </geometry>
    </visual>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <collision name="collision">
      <!-- <origin xyz="0 0 0" rpy="0 1.5707 1.5707"/> -->
      <origin rpy="1.5707 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.1" radius="0.151"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
      <mass value="2"/>
      <inertia ixx=".1" ixy="0.0" ixz="0.0" iyy=".1" iyz="0.0" izz=".1"/>
    </inertial>
  </link>
  <joint name="right_wheel_joint" type="continuous">
    <parent link="chassis_link"/>
    <child link="right_wheel"/>
    <origin rpy="0 0 0" xyz="0 -0.3175 0.1510"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
  </joint>
  <!-- Gazebo wheel reference -->
  <gazebo reference="right_wheel">
    <mu1 value="200.0"/>
    <mu2 value="100.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <material>Gazebo/Black</material>
  </gazebo>
  <!-- Sensor Suite -->
  <link name="imu_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length=".024" radius=".029"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length=".024" radius=".029"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <inertial>
      <mass value="0.09"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <joint name="imu_joint" type="fixed">
    <origin rpy="0 0 0" xyz="-0.04 0 0.3"/>
    <parent link="chassis_link"/>
    <child link="imu_link"/>
  </joint>
  <material name="red">
    <color rgba="1 0 0 1"/>
  </material>
  <material name="green">
    <color rgba="0 1 0 1"/>
  </material>
  <material name="blue">
    <color rgba="0 0 1 1"/>
  </material>
  <material name="white">
    <color rgba="1 1 1 1"/>
  </material>
  <!-- this name is required for the point cloud -->
  <link name="stereo0_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
      <mass value="0.1"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <!-- The camera joint -->
  <joint name="stereo0_joint" type="fixed">
    <origin rpy="-2.913 0.025 -2.466" xyz="-0.003 -0.21 1.773"/>
    <parent link="chassis_link"/>
    <child link="stereo0_link"/>
  </joint>
  <!-- The corrected link for gazebo -->
  <link name="stereo0_corrected_link">
    <inertial>
      <mass value="0"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <joint name="stereo0_corrected_joint" type="fixed">
    <origin rpy="1.57079632679 -1.5707963267 0" xyz="0 0 0"/>
    <parent link="stereo0_link"/>
    <child link="stereo0_corrected_link"/>
  </joint>
  <!-- this name is required for the point cloud -->
  <link name="stereo1_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
      <mass value="0.1"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <!-- The camera joint -->
  <joint name="stereo1_joint" type="fixed">
    <origin rpy="-2.882 -0.038 -0.716" xyz="-0.035 0.343 1.756"/>
    <parent link="chassis_link"/>
    <child link="stereo1_link"/>
  </joint>
  <!-- The corrected link for gazebo -->
  <link name="stereo1_corrected_link">
    <inertial>
      <mass value="0"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <joint name="stereo1_corrected_joint" type="fixed">
    <origin rpy="1.57079632679 -1.5707963267 0" xyz="0 0 0"/>
    <parent link="stereo1_link"/>
    <child link="stereo1_corrected_link"/>
  </joint>
  <!-- this name is required for the point cloud -->
  <link name="stereo2_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
      <mass value="0.1"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <!-- The camera joint -->
  <joint name="stereo2_joint" type="fixed">
    <origin rpy="-1.726 -0.046 -1.642" xyz="-0.361 -0.12 1.238"/>
    <parent link="chassis_link"/>
    <child link="stereo2_link"/>
  </joint>
  <!-- The corrected link for gazebo -->
  <link name="stereo2_corrected_link">
    <inertial>
      <mass value="0"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <joint name="stereo2_corrected_joint" type="fixed">
    <origin rpy="1.57079632679 -1.5707963267 0" xyz="0 0 0"/>
    <parent link="stereo2_link"/>
    <child link="stereo2_corrected_link"/>
  </joint>
  <!-- this name is required for the point cloud -->
  <link name="stereo3_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
      <mass value="0.1"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <!-- The camera joint -->
  <joint name="stereo3_joint" type="fixed">
    <origin rpy="0.026 0.635 -3.098" xyz="-0.241 0.046 1.823"/>
    <parent link="chassis_link"/>
    <child link="stereo3_link"/>
  </joint>
  <!-- The corrected link for gazebo -->
  <link name="stereo3_corrected_link">
    <inertial>
      <mass value="0"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <joint name="stereo3_corrected_joint" type="fixed">
    <origin rpy="1.57079632679 -1.5707963267 0" xyz="0 0 0"/>
    <parent link="stereo3_link"/>
    <child link="stereo3_corrected_link"/>
  </joint>
  <!-- this name is required for the point cloud -->
  <link name="color0_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
      <mass value="0.1"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <!-- The camera joint -->
  <joint name="color0_joint" type="fixed">
    <origin rpy="-2.011 0.033 -1.557" xyz="0.0 -0.019 1.71"/>
    <parent link="chassis_link"/>
    <child link="color0_link"/>
  </joint>
  <!-- The corrected link for gazebo -->
  <link name="color0_corrected_link">
    <inertial>
      <mass value="0"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <joint name="color0_corrected_joint" type="fixed">
    <origin rpy="1.57079632679 -1.5707963267 0" xyz="0 0 0"/>
    <parent link="color0_link"/>
    <child link="color0_corrected_link"/>
  </joint>
  <!-- Standard ZED stuff -->
  <link name="zed_left_camera">
    <visual>
      <origin rpy="0 1.57 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length=".031" radius="0.007"/>
      </geometry>
      <material name="dark_grey">
        <color rgba="0.1 0.1 0.1 1"/>
      </material>
    </visual>
  </link>
  <link name="zed_center">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://zed_wrapper/urdf/ZED.stl"/>
      </geometry>
      <material name="light_grey">
        <color rgba="0.8 0.8 0.8 0.8"/>
      </material>
    </visual>
  </link>
  <link name="zed_right_camera">
    <visual>
      <origin rpy="0 1.57 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length=".031" radius="0.007"/>
      </geometry>
      <material name="dark_grey">
        <color rgba="0.1 0.1 0.1 1"/>
      </material>
    </visual>
  </link>
  <link name="zed_depth_camera"/>
  <joint name="zed_left_camera_joint" type="fixed">
    <parent link="zed_center"/>
    <child link="zed_left_camera"/>
    <origin rpy="0 0 0" xyz="0 0.06 0"/>
  </joint>
  <joint name="zed_depth_camera_joint" type="fixed">
    <parent link="zed_left_camera"/>
    <child link="zed_depth_camera"/>
    <origin rpy="-1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0"/>
  </joint>
  <joint name="zed_right_camera_joint" type="fixed">
    <parent link="zed_center"/>
    <child link="zed_right_camera"/>
    <origin rpy="0 0 0" xyz="0 -0.06 0"/>
  </joint>
  <!-- Addon joint which bolts the zed stuff to my urdf -->
  <joint name="stereo3_zed_addon_joint" type="fixed">
    <parent link="stereo3_link"/>
    <child link="zed_center"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <!-- The link for darknet renamed with the right frame id -->
  <link name="detection_image">
    <inertial>
      <mass value="0"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <joint name="color0_yolo" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="color0_link"/>
    <child link="detection_image"/>
  </joint>
</robot>

